### DISCUSSION TITLE
Skyblock auctions duplicate gemstones issue

### DISCUSSION THREAD

**Discussion by MattTheCuber** [HypixelDev/PublicAPI#549] [Q&A] [answered]
### Hello,

**Problem Statement:** There is no way to reliably find what gemstones an item has applied to it.

I finally decided to dive deep into how gemstones display in the API on items. I ran a lot of logging to try to accumulate what it looks like in the API. I do not currently know if this is limited to the auctions endpoint, I assume it would apply to items on a general level therefore affecting every endpoint that includes items.

### Here is a sample of the possible options:
```json
{
    "TYPE_X": "TIER",
    // and/or
    "CATEGORY_X": "TIER",
    "CATEGORY_X_gem": "Type",
    // and sometimes
    "unlocked_slots": {
        "values": ["TYPE_X", "CATEGORY_X"]
    }
}
```

This looks like it should work fine, that's when you start getting duplicates.
For example auction `9e64458f6dee43839b5d6f67a2fa3e59` has it's gemstones like this:
```json
{
    "COMBAT_0": "FINE",
    "COMBAT_0_gem": "SAPPHIRE",
    "UNIVERSAL_0": "FLAWLESS",
    "UNIVERSAL_0_gem": "SAPPHIRE",
    "SAPPHIRE_0": "FINE"
}
```
Now from in-game this is very easy to see what it has:
![image](https://user-images.githubusercontent.com/32849887/175561904-3a1ad5ce-bf09-487d-8a51-48f49959ce1a.png)
But through looking in the API only, I would assume it had a **Fine Sapphire**, a **Flawless Sapphire**, and another **Fine Sapphire**.

You might think I would be able to ignore keys starting with a gemstone type if there are keys starting with a slot type (e.g. ignore the `SAPPHIRE_0` key), but this is not the case. See the example below of auction `e818236f5c9a42109947f1ef5ec3ae9c`:
```json
{
    "COMBAT_0": "FLAWLESS",
    "COMBAT_0_gem": "SAPPHIRE",
    "SAPPHIRE_0": "FLAWLESS"
}
```
and in-game: 
![image](https://user-images.githubusercontent.com/32849887/175562460-2117ecc6-b55e-4404-a34a-f1b3579e44be.png)

I hope by now you can see the confusion I am having. There is no way to reliably find what gemstones an item has applied to it. ~~I have also tried using the `items` endpoint to find the available slots for an item, but this does not work because you still can't tell what the duplicate gemstone is.~~

I don't know what the cause of this issue is, but I am hoping that the development team will take a look at this issue so that developers can reliably calculate how many and what type of gemstones are applied to items. This issue can cause a large issue on networth calculations, displaying items, listing the gemstones that are applied to an auction.

**Comment by MattTheCuber** [comment, marked as answer]
Found an answer after examining this problem to death. The `/skyblock/items` endpoint comes to the rescue! Basically, instead of looping through the gemstones on an item, you need to loop through the available slots of the item ignoring duplicate keys for the actual item.

This is an example of an item that allows gemstones in the `/skyblock/items` endpoint. After comparing this to the data for an item itself, it is easy to programmatically find what gemstones an item has applied and what slots are unlocked.
```json
"gemstone_slots": [
  {
    "slot_type": "TYPE"
  },
  {
    "slot_type": "TYPE",
    "costs": [...]
  }
]```

**Reply by Semisol** [reply]
You could have ignored the universal slot too.
